File: tutorials/tutorial-java/src/test/java/uitest/pageobjects/DetailPagePo.java
// Ranorex Webtestit Page Object File

package uitest.pageobjects;

import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.*;

public class DetailPagePo {

    // Additional data:
    // {"img":"screenshots/affdea54-ae77-feaf-764d-6e592354e962.png"}
    private By productName = By.cssSelector(".product_title");
    // Additional data:
    // {"img":"screenshots/2f35ac78-865e-947f-15c7-16fb0313e369.png"}
    private By productPrice = By.cssSelector(".entry-summary .woocommerce-Price-amount");
    // Additional data:
    // {"img":"screenshots/5a5345ce-553b-993e-82d9-e22d3dca2307.png"}
    private By addToCartButton = By.cssSelector(".single_add_to_cart_button");
    // Additional data:
    // {"img":"screenshots/7cce2b54-65cf-493b-9c60-b1cdb2ba9998.png"}
    private By cartConfirmationMessage = By.cssSelector("[role='alert']");

    protected WebDriver driver;
    protected WebDriverWait wait;

    private void waitForVisibility(WebElement element) throws Error {
        new WebDriverWait(driver, 5).until(ExpectedConditions.visibilityOf(element));
    }

    public DetailPagePo(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, 10);
    }

    public DetailPagePo open(String url) {
        this.driver.get(url);

        return this;
    }

    public String getTitle() {
        return this.driver.getTitle();
    }

    public String getProductName() {
        return this.wait.until(ExpectedConditions.visibilityOfElementLocated(this.productName)).getText();
    }

    public String getProductPrice() {
        return this.wait.until(ExpectedConditions.visibilityOfElementLocated(this.productPrice)).getText();
    }

    public void addProductToCart() {
        this.wait.until(ExpectedConditions.visibilityOfElementLocated(this.addToCartButton)).click();
    }

    public String getConfirmationMessage() {
        return this.wait.until(ExpectedConditions.visibilityOfElementLocated(this.cartConfirmationMessage)).getText();
    }

}

File: tutorials/tutorial-ts/test/pageobjects/DetailPagePo.ts
// Ranorex Webtestit Page Object File

import { browser, element, by, ExpectedConditions } from 'protractor';

export class DetailPagePo {
  // Additional data: {"img":"screenshots/5140b065-5229-0f94-e96d-d8a41f6acac8.png"}
  private get productName() { return by.css('.product_title'); }
  // Additional data: {"img":"screenshots/de44e8ba-6ff1-4901-0805-871a27c02493.png"}
  private get productPrice() { return by.css('.entry-summary .woocommerce-Price-amount'); }
  // Additional data: {"img":"screenshots/5f1c462d-9e39-5b82-822b-7ff9e1a829c9.png"}
  private get addToCartButton() { return by.css('.single_add_to_cart_button'); }
  // Additional data: {"img":"screenshots/5a23ef9e-c4e1-6d2b-55a6-5c27cca17d06.png"}
  private get cartConfirmationMessage() { return by.css('[role=\'alert\']'); }

  public async open(url: string): Promise<DetailPagePo> {
    await browser.get(url);

    return this;
  }

  public async getTitle(): Promise<string> {
    return await browser.getTitle();
  }

  public async getProductName(): Promise<string> {
    await browser.wait(ExpectedConditions.visibilityOf(element(this.productName)), browser.allScriptsTimeout, this.productName.toString());
    return await element(this.productName).getText();
  }

  public async getProductPrice(): Promise<string> {
    await browser.wait(ExpectedConditions.visibilityOf(element(this.productPrice)), browser.allScriptsTimeout, this.productPrice.toString());
    return await element(this.productPrice).getText();
  }

  public async addProductToCart(): Promise<DetailPagePo> {
    await browser.wait(ExpectedConditions.visibilityOf(element(this.addToCartButton)), browser.allScriptsTimeout, this.addToCartButton.toString());
    await element(this.addToCartButton).click();

    return this;
  }

  public async getConfirmationMessage(): Promise<string> {
    await browser.wait(ExpectedConditions.visibilityOf(element(this.cartConfirmationMessage)), browser.allScriptsTimeout, this.cartConfirmationMessage.toString());
    return await element(this.cartConfirmationMessage).getText();
  }
}
